File: lib/perfchart.pl
/*  Part of SWI-Prolog

    Author:        Jan Wielemaker
    E-mail:        J.Wielemaker@vu.nl
    WWW:           http://www.swi-prolog.org
    Copyright (c)  2019-2020, VU University Amsterdam
                              CWI, Amsterdam
    All rights reserved.

    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions
    are met:

    1. Redistributions of source code must retain the above copyright
       notice, this list of conditions and the following disclaimer.

    2. Redistributions in binary form must reproduce the above copyright
       notice, this list of conditions and the following disclaimer in
       the documentation and/or other materials provided with the
       distribution.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
    COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
    POSSIBILITY OF SUCH DAMAGE.
*/

:- module(webstat_perfchart, []).
:- use_module(library(http/http_dispatch)).
:- use_module(library(http/http_parameters)).
:- use_module(library(http/http_json)).
:- use_module(library(broadcast)).
:- use_module(library(error)).
:- use_module(library(apply)).
:- if(exists_source(library(mallocinfo))).
:- use_module(library(mallocinfo)).
:- endif.
:- if(exists_directory('/proc/self/fd')).
:- use_module(procps).
:- endif.

:- use_module(util).
:- use_module(push).

:- multifile
    webstat:stat_series/2,
    webstat:stat_value/2,
    webstat:stat_disabled/1.

:- http_handler(webstat_api('perf/sample'), perf_sample, [id(perf_sample)]).
:- http_handler(webstat_api('perf/series'), perf_series, [id(perf_series)]).

%!  perf_series(+Request)
%
%   Return the defined performance series. This is a JSON list providing
%   objects with minimally a `name` field.

perf_series(_Request) :-
    findall(Series, active_stat_series(_, Series), SeriesList),
    reply_json(json{ series:SeriesList,
                     rate:1,
                     refresh:1
                   }).

active_stat_series(Name, Dict) :-
    webstat:stat_series(Name, Props),
    \+ webstat:stat_disabled(Name),
    Dict = Props.put(name, Name).
active_stat_series(Name, Dict) :-
    stat_series(Name, Props),
    \+ webstat:stat_disabled(Name),
    Dict = Props.put(name, Name).


		 /*******************************
		 *           MARKINGS		*
		 *******************************/

:- listen(webstat(perfchart, Action),
          wstat(Action)).

wstat(Action) :-
    must_be(ground, Action),
    (   is_list(Action)
    ->  maplist(action_json, Action, JSON)
    ;   action_json(Action, JSON0),
        JSON = [JSON0]
    ),
    webstat_push(JSON).

action_json(marking(Options),
            json{target:perfchart,
                 marking: Obj
                }) :-
    !,
    dict_create(Obj0, json, Options),
    format_label(Obj0, Obj).
action_json(interval(Time),
            json{target:perfchart,
                 interval:Time
                }) :-
    !,
    must_be(number, Time).
action_json(refresh(Time),
            json{target:perfchart,
                 refresh:Time
                }) :-
    !,
    must_be(number, Time).
action_json(series(Series),
            json{target:perfchart,
                 series:JSON
                }) :-
    must_be(list, Series),
    convlist(series_json, Series, JSON).
action_json(Action,
            json{target:perfchart,
                 action:Action
                }) :-
    !,
    must_be(oneof([stop,start,clear]), Action).


format_label(Options0, Options) :-
    Fmt-Args = Options0.get(label),
    !,
    format(string(Label), Fmt, Args),
    Options = Options0.put(label, Label).
format_label(Options0, Options) :-
    Label0 = Options0.get(label),
    \+ (string(Label0); atom(Label0)),
    !,
    term_string(Label0, Label),
    Options = Options0.put(label, Label).
format_label(Options, Options).

series_json(+Series, json{series:Series, show:true}) :-
    stat_series(Series, _),
    !.
series_json(-Series, json{series:Series, show:false}) :-
    stat_series(Series, _),
    !.
series_json(Term, _) :-
    domain_error(series, Term).


		 /*******************************
		 *            SAMPLING		*
		 *******************************/

%!  perf_sample(+Request)
%
%   Return a JSON object providing values for each defined series.

perf_sample(Request) :-
    http_parameters(Request,
                    [ thread(Thread, [default(main)])
                    ]),
    in_thread(Thread, perf_sample_data(Data)),
    reply_json(json{sample:Data}).

:- discontiguous
    stat_series/2,
    stat/2.

stat_series(global,
            _{ label: "Global stack",
               unit:  bytes
             }).
stat_series(trail,
            _{ label: "Trail stack",
               unit:  bytes
             }).
stat_series(local,
            _{ label: "Local stack",
               unit:  bytes
             }).
stat_series(stack,
            _{ label: "Total stack usage",
               unit:  bytes,
               active: true
             }).

stat_series(thread_cpu,
            _{ label: "Thread CPU",
               unit:  percent,
               zero_ok: true
             }).
stat_series(process_cpu,
            _{ label: "Process CPU",
               unit:  percent,
               zero_ok: true,
               active: true
             }).

stat_series(atoms,
            _{ label: "Atoms"
             }).
stat_series(functors,
            _{ label: "Functors"
             }).

stat_series(modules,
            _{ label: "Modules"
             }).
stat_series(predicates,
            _{ label: "Predicates"
             }).
stat_series(clauses,
            _{ label: "clauses"
             }).
stat_series(code_mem,
            _{ label: "Memory for VM code",
               unit:  bytes
             }).

stat_series(tables,
            _{ label: "Tables"
             }).
stat_series(table_space,
            _{ label: "Table space",
               unit:  bytes,
               active: true
             }).

perf_sample_data(Data) :-
    findall(Name-Value, active_stat(Name, Value), Pairs),
    dict_create(Data, stat, Pairs).

active_stat(Name, Value) :-
    (   stat(Name, Value),
        once(stat_series(Name, _))
    ;   webstat:stat_value(Name, Value),
        once(webstat:stat_series(Name, _))
    ),
    \+ webstat:stat_disabled(Name).

:- thread_local
    prev_sample/3.

stat(thread_cpu, Value) :-
    \+ webstat:stat_disabled(thread_cpu),
    get_time(Wall1),
    statistics(cputime, CPU1),
    debug(profiler, 'stat(thread_cpu, ~p)', [CPU1]),
    (   retract(prev_sample(thread_cputime, Wall0, CPU0))
    ->  asserta(prev_sample(thread_cputime, Wall1, CPU1)),
        Value is round(100*(CPU1-CPU0)/(Wall1-Wall0))
    ;   asserta(prev_sample(thread_cputime, Wall1, CPU1)),
        Value is 0
    ).
stat(process_cpu, Value) :-
    \+ webstat:stat_disabled(process_cpu),
    get_time(Wall1),
    statistics(process_cputime, CPU1),
    (   retract(prev_sample(process_cputime, Wall0, CPU0))
    ->  asserta(prev_sample(process_cputime, Wall1, CPU1)),
        Value is round(100*(CPU1-CPU0)/(Wall1-Wall0))
    ;   asserta(prev_sample(process_cputime, Wall1, CPU1)),
        Value = 0
    ).


stat(global, Value) :-
    statistics(globalused, Value).
stat(trail, Value) :-
    statistics(trailused, Value).
stat(local, Value) :-
    statistics(localused, Value).
stat(stack, Value) :-
    statistics(stack, Value).

stat(atoms, Value) :-
    statistics(atoms, Value).
stat(functors, Value) :-
    statistics(functors, Value).

stat(modules, Value) :-
    statistics(modules, Value).
stat(predicates, Value) :-
    statistics(predicates, Value).
stat(clauses, Value) :-
    statistics(clauses, Value).
stat(code_mem, Bytes) :-
    statistics(codes, Codes),
    current_prolog_flag(address_bits, Bits),
    Bytes is Codes*(Bits/8).

stat(tables, Value) :-
    \+ webstat:stat_disabled(tables),
    aggregate_all(sum(C), ('$tbl_variant_table'(VariantTrie),
                           trie_property(VariantTrie, value_count(C))),
                  Value).
stat(table_space, Value) :-
     statistics(table_space_used, Value).

:- if(current_predicate(malloc_property/1)).
stat(tcheap, Heap) :-
    statistics(heapused, Heap).

stat(tcheap_fragmentation, Lost) :-
    malloc_property('generic.heap_size'(Total)),
    malloc_property('generic.current_allocated_bytes'(Allocated)),
    Lost is Total-Allocated.

stat_series(tcheap,
            _{ label: "Heap memory",
               title: "Heap (malloc - stack)",
               unit:  bytes
             }).
stat_series(tcheap_fragmentation,
            _{ label: "Malloc lost",
               title: "Freed memory not reused",
               unit:  bytes
             }).
:- endif.

:- if(current_predicate(mallinfo/1)).
stat_series(malloc,
            _{ label: "Malloc in use",
               title: "Malloc'ed memory in use",
               unit:  bytes
             }) :-
    \+ webstat:stat_disabled(mallinfo).
stat_series(mfree,
            _{ label: "Malloc free",
               title: "Freed memory not reused",
               unit:  bytes
             }) :-
    \+ webstat:stat_disabled(mallinfo).
:- if(current_predicate(procps_stat/1)).
stat_series(rss,
            _{ label: "RSS memory",
               title: "Resident Set Size",
               unit:  bytes
             }) :-
    \+ webstat:stat_disabled(procps).
stat_series(heap,
            _{ label: "Heap memory",
               title: "RSS - stacks - free",
               unit:  bytes
             }) :-
    \+ webstat:stat_disabled(mallinfo),
    \+ webstat:stat_disabled(procps).
:- endif.

% Overall memory usage stats. Only for Linux.  Note that mallinfo() is a
% GNU extension and does not report  values   larger  than  4Gb. It also
% tends to get really slow if  the   amount  of allocated memory becomes
% large. Add the line below to your   project  to disable using mallinfo
% based stats.
%
%     webstat:stat_disabled(mallinfo)
%
% RSS is extracted from the Linux /proc  file system and can be disabled
% using
%
%     webstat:stat_disabled(procps)

stat(Stat, Value) :-
    (   (   webstat:stat_disabled(malloc),
            webstat:stat_disabled(mfree),
            webstat:stat_disabled(heap)
        ;   webstat:stat_disabled(mallinfo)
        )
    ->  (   (   webstat:stat_disabled(rss)
            ;   webstat:stat_disabled(procps)
            )
        ->  fail
        ;   procps_stat(ProcPS),
            Stat = rss,
            Value is ProcPS.rss
        )
    ;   mallinfo(Dict),
        mallinfo_stat(Stat, Dict, Value)
    ).

mallinfo_stat(malloc, Dict, Value) :-
    Value = Dict.uordblks.
mallinfo_stat(mfree, Dict, Value) :-
    Value = Dict.fordblks.
:- if(current_predicate(procps_stat/1)).
mallinfo_stat(Stat, Mallinfo, Value) :-
    \+ webstat:stat_disabled(procps),
    procps_stat(ProcPS),
    procps_mallinfo_stat(Stat, Mallinfo, ProcPS, Value).

procps_mallinfo_stat(heap, Mallinfo, ProcPS, Value) :-
    statistics(stack, Stack),
    Value is ProcPS.rss - Mallinfo.fordblks - Stack.
procps_mallinfo_stat(rss, _Mallinfo, ProcPS, Value) :-
    Value is ProcPS.rss.
:- else.
procps_stat(_) :- fail.
:- endif.
:- endif.

File: web/js/perfchart.js
/*  Part of SWISH

    Author:        Jan Wielemaker
    E-mail:        J.Wielemaker@cs.vu.nl
    WWW:           http://www.swi-prolog.org
    Copyright (C): 2019-2020, VU University Amsterdam
			      CWI Amsterdam
    All rights reserved.

    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions
    are met:

    1. Redistributions of source code must retain the above copyright
       notice, this list of conditions and the following disclaimer.

    2. Redistributions in binary form must reproduce the above copyright
       notice, this list of conditions and the following disclaimer in
       the documentation and/or other materials provided with the
       distribution.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
    COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
    POSSIBILITY OF SUCH DAMAGE.
*/

/**
 * @fileOverview
 * Dynamic performance charts for monitoring memory usage, etc.
 *
 * @version 0.2.0
 * @author Jan Wielemaker, J.Wielemaker@vu.nl
 * @requires jquery
 */

define([ "jquery", "config", "flot", "utils", "form", "modal",
	 "palette", "laconic" ],
       function($, config, plot, utils, form, modal) {

(function($) {
  var pluginName = 'perfchart';

  var plot_default_options = {
    xaxis: {
      min: 0,
      max: 1000
    },
    yaxis: {
      min: 1,
      max: 1000000000000,
      ticks: [ 1,10,100,1000,10000,100000,
	       1000000,10000000,100000000,1000000000,
	       10000000000, 100000000000, 1000000000000
	     ],
      tickFormatter: suffixFormatter,
      transform: function(v) {return Math.log(v+1);}
    },
    grid: {
      hoverable: true
    }
  };

  var default_refresh_rate = 1.0;	/* refresh max each second */

  /** @lends $.fn.perfchart */
  var methods = {
    _init: function(options) {
      return this.each(function() {
	var elem = $(this);
	var data = {};			/* private data */
	var hsplit;

	elem.data(pluginName, data);	/* store with element */
	elem.addClass("perfchart reactive-size swish-event-receiver listen-close-event");
	elem[pluginName]('controller');
	elem.append(hsplit=$($.el.div({class: "hsplit_lp"})));
	elem[pluginName]('series_menu');
	hsplit.append($.el.div({class:"flot"}));
	elem[pluginName]('tooltip');
	elem.on('reactive-resize', function() {
	  elem[pluginName]('resize');
	});
	elem.on('tab-close', function() {
	  if ( data.timer ) {
	    clearInterval(data.timer);
	    delete(data.timer);
	  }
	});
	elem.on('activate-tab', function() {
	  elem[pluginName]('redraw');
	});

	$.ajax({
	  url: config.http.locations.perf_series,
	  success: function(data) {
	    elem[pluginName]('series', data);
	  },
	  error: function(jqXHDR) {
	    modal.ajaxError(jqXHDR);
	  }
	});
      })
    },

    /**
     * Initialize the data series and flot instance from the reported
     * available series.
     */
    series: function(options) {
      var elem = $(this);
      var data = elem.data(pluginName);
      var colors = palette("mpn65", options.series.length);

      data.rate      = options.rate||1;
      data.refresh   = options.refresh||1;
      data.series    = {};
      data.x	     = 0;
      data.samples   = Math.round(elem.width()/2);
      data.flot_data = [];
      data.flot_opts = $.extend(true, {}, plot_default_options);

      data.flot_opts.xaxis.max = data.flot_opts.xaxis.min + data.samples;

      for(var i=0; i<options.series.length; i++) {
	var series = options.series[i];

	series.data  = [];
	if ( !series.label )
	  series.label = series.name;
	if ( !series.color )
	  series.color = "#"+colors[i];

	data.series[series.name] = series;
	if ( series.active ) {
	  var idx = data.flot_data.length;
	  series.index = idx;
	  data.flot_data.push(series);
	}
      }

      elem[pluginName]('series_menu');

      data.plot = $.plot(elem.find(".flot"), data.flot_data, data.flot_opts);
      data.plot.setupGrid();
      elem[pluginName]('update');
      elem[pluginName]('play', true);
    },

    activate_series: function(name, val) {
      var elem   = $(this);
      var data   = $(this).data(pluginName);
      var series = data.series[name];

      if ( val ) {
	var idx = data.flot_data.length;
	series.index = idx;
	data.flot_data.push(series);
      } else {
	data.flot_data.splice(series.index, 1);
	for(var i=0; i<data.flot_data.length; i++)
	  data.flot_data[i].index = i;
      }

      data.plot.setData(data.flot_data);
      data.plot.setupGrid();
      data.plot.draw();
    },

    /**
     * Deal with push messages
     */
    push: function(obj) {
      var elem = $(this);
      var data  = $(this).data(pluginName);

      if ( obj.marking ) {
	var marking = obj.marking;
	marking.x = data.x;
	this[pluginName]('marking', marking);
      }
      if ( obj.action ) {
	if ( obj.action == 'stop' )
	  elem[pluginName]('play', false);
	else if ( obj.action == 'start' )
	  elem[pluginName]('play', true);
	else if ( obj.action == 'clear' )
	  elem[pluginName]('clear');
      }
      if ( obj.interval ) {
	elem[pluginName]('play', obj.interval);
      }
      if ( obj.refresh ) {
	if ( obj.refresh < data.refresh )
	  elem[pluginName]('redraw');
	data.refresh = obj.refresh;
      }
      if ( obj.series ) {
	var menu = elem.find("div.series");

	for(var i=0; i<obj.series.length; i++) {
	  var s = obj.series[i];

	  elem[pluginName]('activate_series', s.series, s.show);
	  var ch = menu.find("input[name="+s.series+"]");
	  if ( s.show )
	    ch.attr("checked", "checked");
	  else
	    ch.removeAttr("checked");
	}
      }
    },

    /**
     * Add a marking annotation to the X-axis
     */
    marking: function(options, draw) {
      var data  = $(this).data(pluginName);
      var color = options.color||"#000";
      var lw    = options.width||1;
      var x     = options.x||0;
      var y     = options.y||11;
      var grid  = data.plot.getOptions().grid;

      if ( grid.markings == undefined )
	grid.markings = [];

      grid.markings.push({
        color: color,
	lineWidth: lw,
	xaxis: {from: x, to: x}
      });

      if ( options.label ) {
	var placeholder = $(this).find(".flot");
	var o = data.plot.pointOffset({ x: x, y: Math.pow(10,y)});
	var m;
	placeholder.append(m=
			   $.el.div({style: "position:absolute;"+
					    "left:"+(o.left+4)+"px;"+
					    "top:"+(o.top)+"px;",
				     class: "plot-marking-label"
				    },
				    options.label));
	$(m).data("marking", {x:x, y:y});
      }

      if ( draw != false ) {
	data.plot.setupGrid();
	data.plot.draw();
      }
    },

    /* Update the marking labels.  If the marking is outside the xaxis
     * region, delete it completely.
     */

    update_markings: function() {
      var elem  = $(this);
      var data  = elem.data(pluginName);
      var xaxis = data.plot.getAxes().xaxis;

      elem.find(".plot-marking-label").each(function() {
	var div = $(this);
	var md  = div.data("marking");

	if ( md.x < xaxis.options.min || md.x > xaxis.options.max ) {
	  md.remove();
	} else {
	  var o = data.plot.pointOffset({ x: md.x, y: Math.pow(10,md.y)});
	  div.css({ left: o.left+4, top: o.top });
	}
      });
    },

    /**
     * Start collecting the chart.
     * @param {Any} how is one of `false`, `true` or #milliseconds
     */
    play: function(how) {
      var elem  = $(this);
      var data  = $(this).data(pluginName);
      var clear = (how == false);

      if ( typeof how == "number" ) {
	if ( data.rate != how ) {
	  data.rate = how;
	  if ( data.timer ) {
	    how = true;
	    clear = true;
	  } else {
	    how = false;
	  }
	}
      }

      if ( clear && data.timer ) {
	clearInterval(data.timer);
	data.timer = null;
	if ( !how )			/* stopped, flush last bits */
	  elem[pluginName]('redraw');
      }

      if ( how == true && !data.timer ) {
	data.timer = setInterval(function() {
	  elem[pluginName]('update');
	}, Math.round(data.rate*1000));
      }

      var btn = elem.find(".controller .play");
      var span = btn.find("span");
      span.removeClass("glyphicon-play glyphicon-pause");
      if ( how ) {
	btn.addClass("recording")
	   .attr("title", "Stop recording");
	span.addClass("glyphicon-pause");
      } else {
	btn.removeClass("recording")
	   .attr("title", "Start recurding");
	span.addClass("glyphicon-play");
      }

      var sel = elem.find(".controller select");
      sel.find("option").removeAttr("selected");
      var opt = sel.find("option[value='"+data.rate+"']");
      if ( opt.length > 0 ) {
	opt.attr("selected", "selected");
      } else {
	sel.append($.el.option({value:data.rate, selected:"selected"},
			       data.rate+" sec"));
      }
    },

    controller: function() {
      var elem = $(this);
      var data = $(this).data(pluginName);
      var ctrl;
      var br;
      var play;
      var pause, clear;

      elem.append(ctrl=$($.el.div({class:"form-inline controller"})));
      ctrl.append(br=$($.el.div({class:"btn-group"})));
      br.append(clear=$(form.widgets.glyphIconButton("step-backward", {})),
		$("<span class='menu-space'>&nbsp</span>"),
		play =$(form.widgets.glyphIconButton("play", {})));

      clear.on("click", function() { elem[pluginName]('clear'); })
           .attr("title", "Clear and reset chart");
      play.addClass("play");
      play.on("click", function(ev) {
	elem[pluginName]('play', data.timer ? false : true);
      });

      br.append($.el.label({class:"sample-rate"}, "Sample rate:"));
      br.append(sel=$($.el.select({class:"form-control"})));

      sel.on('change', function() {
	elem[pluginName]('play', this.value);
      });

      function opt(time, label, def) {
	var opts = {"value":time};
	if ( def ) opts.selected = "selected";

	sel.append($.el.option(opts, label));
      }

      opt(0.2, "0.2 sec");
      opt(0.5, "0.5 sec");
      opt(1.0, "1 sec", true);
      opt(2.0, "2 sec");
      opt(5.0, "5 sec");
      opt(10.0, "10 sec");
      opt(60.0, "1 min");
      opt(120.0, "2 min");
      opt(300.0, "5 min");
    },

    series_menu: function() {
      var elem = $(this);
      var data = $(this).data(pluginName);
      var lbl = $.el.label("Series");
      var menu;

      if ( (menu=elem.find("div.series")) && menu.length > 0 ) {
	menu.empty().append(lbl);
      } else {
	elem.find(".hsplit_lp").append(menu=$($.el.div({class:"series"}, lbl)));
	menu.on("change", "input", function(ev) {
	  var name    = $(ev.target).attr('name');
	  var checked = $(ev.target).prop('checked') == true;

	  elem[pluginName]('activate_series', name, checked);
	});
      }

      if ( data.series ) {
	for(var p in data.series) {
	  if ( data.series.hasOwnProperty(p) ) {
	    var series = data.series[p];

	    menu.append(form.widgets.checkbox(series.name,
					      { checked: series.active,
						color: series.color,
						label: series.label,
						title: series.title
					      }));
	  }
	}
      }
    },

    /**
     * Enable the tooltip to get precise values
     */
    tooltip: function() {
      var data    = $(this).data(pluginName);
      var plotdiv = $(this).find(".flot");

      $(plotdiv).hover(function() { $("#flot-tooltip").show(); },
		       function() { $("#flot-tooltip").hide(); });
      $(plotdiv).on("plothover", function (event, pos, item) {
	var x = Math.round(pos.x);
	var t  = $.el.table();
	var hasval = 0;

	function val(ar, x) {
	  if ( ar[0] ) {
	    var x0 = ar[0][0];
	    var x1 = x-x0;
	    var mx = 5;

	    while(!(ar[x1] && ar[x1][0] == x) && --mx >= 0) {
	      if ( x1 >= ar.length )
		x1 = ar.length-1;
	      else if ( x1 < 0 )
		x1 = 0;
	      else
		x1 += x-ar[x1][0];
	    }

	    if ( mx > 0 )
	      return ar[x1][1];
	  }
	}

	$("#flot-tooltip").empty().append(t);

	for(var i=0; i<data.flot_data.length; i++) {
	  var series = data.flot_data[i];
	  var value, str;

	  if ( (value = val(series.data, x)) != undefined ) {
	    if ( series.unit == 'bytes' )
	      str = utils.human_size(value);
	    else if ( series.unit == 'percent' )
	      str = value + "%";
	    else
	      str = utils.human_count(value);
	    hasval++;
	  } else {
	    str = "n/a";
	  }

	  t.append($.el.tr($.el.th(series.label),
			   $.el.td(str)));
	}

	if ( hasval > 0 ) {
	  $("#flot-tooltip")
	      .css({top: pos.pageY+5, left: pos.pageX+5})
	      .show();
	} else {
	  $("#flot-tooltip").hide();
	}
      });
    },

    update: function() {
      var elem = $(this);

      $.ajax({
        url: config.http.locations.perf_sample,
	success: function(data) {
	  elem[pluginName]('add_sample', data);
	},
	error: function(jqXHDR) {
	  elem[pluginName]('play', false);	/* stop on error */
	  modal.ajaxError(jqXHDR);
	}
      });
    },

    /**
     * Add an individual sample
     */
    add_sample: function(options) {
      var elem   = $(this);
      var data   = elem.data(pluginName);
      var sample = options.sample||{};
      var x      = data.x++;
      var xaxis  = data.plot.getAxes().xaxis;
      var rm;

      if ( x > xaxis.options.max ) {
	rm = x - xaxis.options.max;
	xaxis.options.max += rm;
	xaxis.options.min += rm;
      } else
	rm = 0;

      for(var p in data.series) {
	if ( data.series.hasOwnProperty(p) ) {
	  var series = data.series[p];

	  if ( sample[p] || series.zero_ok ) {
	    if ( rm )
	      series.data.splice(0, rm);
	    series.data.push([x, sample[p]]);
	  }
	}
      }

      if ( options.markings ) {
	for(var i=0; i < options.markings.length; i++) {
	  var m = options.markings[i];
	  m.x = x;

	  elem[pluginName]('marking', m, false);
	}
      }

      if ( elem.is(":visible") ) {
	var now = Date.now();
	if ( !(data.last_redraw && now - data.last_redraw < data.refresh*1000) ) {
	  data.last_redraw = now;
	  data.plot.setData(data.flot_data);
	  data.plot.setupGrid();		/* remove if grid is fixed */
	  data.plot.draw();
	  elem[pluginName]('update_markings');
	}
      }
    },

    redraw: function() {
      var elem   = $(this);
      var data   = elem.data(pluginName);

      if ( elem.is(":visible") ) {
	data.plot.setData(data.flot_data);
	data.plot.resize();
	data.plot.setupGrid();			/* remove if grid is fixed */
	data.plot.draw();
	elem[pluginName]('update_markings');
      }
    },

    clear: function() {
      var elem  = $(this);
      var data  = elem.data(pluginName);
      var xaxis = data.plot.getAxes().xaxis;

      for(var p in data.series) {
	if ( data.series.hasOwnProperty(p) ) {
	  var series = data.series[p];
	  series.data.splice(0);
	}
      }

      elem.find(".plot-marking-label").remove();

      data.x = 0;
      xaxis.options.min = data.x;
      xaxis.options.max = data.x + data.samples;

      data.plot.setData(data.flot_data);
      data.plot.getOptions().grid.markings = [];
      data.plot.setupGrid();			/* remove if grid is fixed */
      data.plot.draw();
    },

    resize: function() {
      var elem  = $(this);

      if ( elem.is(":visible") ) {
	var data  = elem.data(pluginName);
	var xaxis = data.plot.getAxes().xaxis;
	var neww  = Math.round(elem.width()/2);
	var min   = xaxis.options.min;

	data.samples = neww;

	if ( neww < data.x ) {
	  var newmin = data.x - neww;
	  var rm = newmin - min;

	  for(var p in data.series) {
	    if ( data.series.hasOwnProperty(p) ) {
	      var series = data.series[p];

	      series.data.splice(0, rm);
	    }
	  }

	  xaxis.options.max = data.x;
	  xaxis.options.min = newmin;
	} else {
	  xaxis.options.max = xaxis.options.min + data.samples;
	}

	data.plot.resize();
	data.plot.setupGrid();
	data.plot.draw();
	elem[pluginName]('update_markings');
      }
    }
  }; // methods

  function suffixFormatter(val, axis) {
    if (val >= 1000000000000)
      return (val / 1000000000000).toFixed(axis.tickDecimals) + " T";
    if (val >= 1000000000)
      return (val / 1000000000).toFixed(axis.tickDecimals) + " G";
    else if (val >= 1000000)
      return (val / 1000000).toFixed(axis.tickDecimals) + " M";
    else if (val >= 1000)
      return (val / 1000).toFixed(axis.tickDecimals) + " K";
    else
      return val.toFixed(axis.tickDecimals) + "  ";
  }

  $("<div id='flot-tooltip'></div>").appendTo("body");

  /**
   * <Class description>
   *
   * @class perfchart
   * @tutorial jquery-doc
   * @memberOf $.fn
   * @param {String|Object} [method] Either a method name or the jQuery
   * plugin initialization object.
   * @param [...] Zero or more arguments passed to the jQuery `method`
   */

  $.fn.perfchart = function(method) {
    if ( methods[method] ) {
      return methods[method]
	.apply(this, Array.prototype.slice.call(arguments, 1));
    } else if ( typeof method === 'object' || !method ) {
      return methods._init.apply(this, arguments);
    } else {
      $.error('Method ' + method + ' does not exist on jQuery.' + pluginName);
    }
  };
}(jQuery));
});
